core/domain/exp_jobs_one_off.py

```diff
@@ -164,120 +248,163 @@ class MathExpressionValidationOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+                    elif len(types_of_inputs) > 1:
+                        yield ('ERROR', (
+                            'The exploration with ID: %s and state name: %s '
+                            'contains inputs that correspond to multiple '
+                            'different types: %s. Please resolve this before '
+                            'running the upgrade job.' % (
+                                item.id, state_name,
+                                ', '.join(list(types_of_inputs)))))
+
+    @staticmethod
+    def reduce(key, values):
+        if key.endswith('Input'):
+            yield (key, values[:VALID_MATH_INPUTS_YIELD_LIMIT])
+        else:
+            yield (key, values)
+
+
+class MathExpressionUpgradeOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of explorations that use the MathExpressionInput
+    along with the validity and type (expression/equation) of the inputs present
+    in the exploration. For valid explorations, they are upgraded to the
+    appropriate new interaction.
+
+    This validation is done by the validator functions present in schema_utils.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        is_valid_algebraic_expression = schema_utils.get_validator(
+            'is_valid_algebraic_expression')
+        is_valid_numeric_expression = schema_utils.get_validator(
+            'is_valid_numeric_expression')
+        is_valid_math_equation = schema_utils.get_validator(
+            'is_valid_math_equation')
+        ltt = latex2text.LatexNodes2Text()
+
+        if not item.deleted:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+            for state_name, state in exploration.states.items():
+                if state.interaction.id == 'MathExpressionInput':
+                    new_answer_groups = []
+                    types_of_inputs = set()
+                    for group in state.interaction.answer_groups:
+                        new_answer_group = copy.deepcopy(group)
+                        for rule_spec in new_answer_group.rule_specs:
+                            rule_input = ltt.latex_to_text(
+                                rule_spec.inputs['x'])
+
+                            rule_input = _clean_math_expression(rule_input)
+
+                            type_of_input = _TYPE_INVALID_EXPRESSION
+                            if is_valid_algebraic_expression(rule_input):
+                                type_of_input = _TYPE_VALID_ALGEBRAIC_EXPRESSION
+                            elif is_valid_numeric_expression(rule_input):
+                                type_of_input = _TYPE_VALID_NUMERIC_EXPRESSION
+                            elif is_valid_math_equation(rule_input):
+                                type_of_input = _TYPE_VALID_MATH_EQUATION
+
+                            types_of_inputs.add(type_of_input)
+
```

Comment:
At this point, shouldn't you be raising and breaking if the expression has invalid type?

You should not be doing any edits at all to an exploration if any "badness" is discovered.

Reply:
We don't wanna break upon finding invalid inputs, because we need to output them so that they can be resolved. And if any of the types_of_inputs is found invalid, we don't modify that state at all. This is checked after the loop on the answer groups is finished.